cli/cmd/report/report.go

```diff
@@ -2,57 +2,94 @@ package report
 
 import (
 	"errors"
-	"fmt"
+	"os"
 	"os/exec"
 	"runtime"
 
+	"numerous/cli/cmd/output"
+
 	"github.com/spf13/cobra"
 )
 
 const numerousReportURL = "https://numerous.com"
 
+var (
+	ErrWSLCheck        = errors.New("error when checking windows subsystem for linux")
+	ErrOSCheck         = errors.New("error to identify the current OS")
+	ErrGrepBashCommand = errors.New("error in the grep bash command execution")
+)
+
 var ReportCmd = &cobra.Command{
 	Use:   "report",
 	Short: "Opens numerous report and feedback page.",
 	Args:  cobra.NoArgs,
 	Run: func(cmd *cobra.Command, args []string) {
-		if err := openURL(numerousReportURL); err != nil {
-			fmt.Println("Error:", err)
+		if err := openURL(numerousReportURL, runtime.GOOS, OSCommandExecutor{}); err != nil {
+			os.Exit(1)
 		}
 	},
 }
 
-func openURL(url string) error {
-	cmd, args, err := setCmdByOS()
+type CommandExecutor interface {
+	Output(command string, args ...string) ([]byte, error)
+	Start(command string, args ...string) error
+}
+
+type OSCommandExecutor struct{}
+
+func (osExec OSCommandExecutor) Output(command string, args ...string) ([]byte, error) {
+	return exec.Command(command, args...).Output()
+}
+
+func (osExec OSCommandExecutor) Start(command string, args ...string) error {
+	return exec.Command(command, args...).Start()
+}
+
+func openURL(url string, os string, exec CommandExecutor) error {
+	cmd, args, err := setCmdByOS(os, exec, isWSL)
 	if err != nil {
+		output.PrintErrorDetails("An error occurred when opening the numerous report page.", err)
 		return err
 	}
-	fmt.Println("Opening the report page in your default browser.")
 	args = append(args, url)
 
-	return exec.Command(cmd, args...).Start()
+	return exec.Start(cmd, args...)
 }
 
-func setCmdByOS() (string, []string, error) {
-	switch runtime.GOOS {
+func setCmdByOS(os string, exec CommandExecutor, wsl func(exec CommandExecutor) (bool, error)) (string, []string, error) {
```

Comment:
```go
func osOrWSL(exec CommandExecutor, os string) string {
    if os == "linux" && isWSL(exec) {
        return "wsl"
    }
    return os
}

```

Reply:
Looks much better this way indeed, but is it not faster to use boolean instead of string (comparison etc)? Probably in a larger scope it will not make a difference right?

Reply:
It is simpler control flow I think to just wrap the OS and add this additional possible value. The string comparison does not really matter in this case, especially since we are comparing strings of a constant size, we still basically have a constant time operation in the comparison :smile: